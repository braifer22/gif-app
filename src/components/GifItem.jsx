import { string } from 'prop-types';

export function GifItem({ title, src, height, width }) {
    return (
        <li>
            <img
                src={src}
                alt={title}
                style={{ aspectRatio: `${width}/${height}` }}
            />
            <p>{title}</p>
        </li>
    );
}

GifItem.propTypes = {
    title: string.isRequired,
    src: string.isRequired,
};
