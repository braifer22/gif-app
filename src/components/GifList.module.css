.title {
    font-size: 2.5rem;
    margin-block: 32px 12px;
    text-transform: capitalize;
}

.gifGrid {
    columns: 3;
    gap: 32px;
    list-style: none;

    li {
        border-radius: 8px;
        outline: 2px solid lightgray;
        outline-offset: 3px;
        overflow: hidden;
        position: relative;
        margin-bottom: 32px;

        &::after {
            content: '';
            position: absolute;
            inset: 0;
            background: linear-gradient(transparent, 65%, rgba(0, 0, 0, 0.829));
        }

    }

    img {
        width: 100%;
        display: block;
    }

    p {
        font-size: 1.2rem;
        padding: 1rem;
        font-weight: 600;
        position: absolute;
        bottom: 0;
        left: 0;
        right: 0;
        z-index: 1;
    }
}

.loading {
    text-align: center;

}