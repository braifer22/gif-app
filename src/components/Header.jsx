import { Search } from './Search.jsx';
import styles from './Header.module.css';
import { useState } from 'react';
import { arrayOf, func, string } from 'prop-types';

export function Header({ onSubmit, categories }) {
    const [value, setValue] = useState('');

    function handleSubmit(event) {
        event.preventDefault();

        if (value.trim() === '') return;
        if (categories.includes(value.trim())) return;

        onSubmit(value);
        setValue('');
    }

    return (
        <header className={styles.header}>
            <form onSubmit={handleSubmit}>
                <label>
                    <Search width={18} strokeColor="rgb(102,102,102)" />
                    <input
                        value={value}
                        onChange={(e) => setValue(e.target.value)}
                        type="text"
                        placeholder="Buscar gifs..."
                    />
                </label>
            </form>
        </header>
    );
}

Header.propTypes = {
    onSubmit: func.isRequired,
    categories: arrayOf(string),
};
