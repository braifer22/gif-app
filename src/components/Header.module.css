.header {
    position: sticky;
    top: 0;
    background-color: Canvas;
    z-index: 1000;

    display: flex;
    gap: 12px;
    justify-content: center;
    align-items: center;
    border-bottom: 1px solid #222;

    form {
        margin: 8px;
    }


    label {
        background-color: #161618;
        border: 2px solid transparent;
        border-radius: 8px;
        display: flex;
        align-items: center;
        padding: 2px 14px;
        transition: .3s ease-out;
        cursor: text;

        &:hover {
            border-color: #5c73e7;
        }
    }

    input {
        background-color: transparent;
        padding: 12px;
        border: none;
        outline: none;
        font-size: 1rem;
    }

    label:has(input:focus) {
        border-color: #5c73e7;
    }
}