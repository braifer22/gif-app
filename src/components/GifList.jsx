import { string } from 'prop-types';
import styles from './GifList.module.css';
import { GifItem } from './GifItem';
import { useFetchGifs } from '../hooks/useFetchGifs';

export function GifList({ category }) {
    const { images, isLoading } = useFetchGifs(category);

    return (
        <article>
            <h1 className={styles.title}>{category}</h1>
            {isLoading ? (
                <p className={styles.loading}>Cargando...</p>
            ) : (
                <>
                    <ul className={styles.gifGrid}>
                        {images.map((image) => {
                            return (
                                <GifItem
                                    key={image.id}
                                    src={image.url}
                                    title={image.title}
                                    height={image.height}
                                    width={image.width}
                                />
                            );
                        })}
                    </ul>
                    {images.length === 0 && (
                        <p>
                            No encontramos Gifs sobre {category} {':('}
                        </p>
                    )}
                </>
            )}
        </article>
    );
}

GifList.propTypes = {
    category: string.isRequired,
};
