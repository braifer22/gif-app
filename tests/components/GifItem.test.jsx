import { render, screen } from '@testing-library/react';
import { describe, expect, test } from 'vitest';
import { GifItem } from '../../src/components/GifItem';

describe('Pruebas <GifItem />', () => {
    const title = 'Alexis';
    const url = 'https://one-punch.com/saitama.jpg';

    test('Debe hacer match con el snapshot', () => {
        const { container } = render(<GifItem title={title} src={url} />);
        expect(container).toMatchSnapshot();
    });

    test('Debe mostrar la imagen con el SRC y el ALT indicado', () => {
        render(<GifItem title={title} src={url} />);
        expect(screen.getByRole('img').getAttribute('src')).toBe(url);
        expect(screen.getByRole('img').getAttribute('alt')).toBe(title);
    });

    test('Debe mostrar el titulo en el componente', () => {
        render(<GifItem title={title} src={url} />);
        expect(screen.getByText(title)).toBeTruthy();
    });
});
