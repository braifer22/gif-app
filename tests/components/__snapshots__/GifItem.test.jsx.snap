// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Pruebas <GifItem /> > Debe hacer match con el snapshot 1`] = `
<div>
  <li>
    <img
      alt="Alexis"
      src="https://one-punch.com/saitama.jpg"
      style="aspect-ratio: undefined/undefined;"
    />
    <p>
      Alexis
    </p>
  </li>
</div>
`;
